src/disk.rs

```diff
@@ -0,0 +1,467 @@
+            PageState::Buffer(_) => {}
+            PageState::Clean(_) => {}
+            PageState::Dirty(i) => {
+                self.driver.write(i, self.caches[0].buf, self.size)?;
+            }
+        }
+
+        Ok(true)
+    }
+
+    pub fn flush(&mut self) -> Result<(), DiskError> {
+        for c in self.caches.iter_mut() {
+            match c.page {
+                PageState::Empty => {}
+                PageState::Buffer(_) => {}
+                PageState::Clean(_) => {}
+                PageState::Dirty(idx) => self.driver.write(idx, c.buf, self.size)?,
+            }
+            c.page = PageState::Empty;
+        }
+        Ok(())
+    }
+
+    pub fn buffer(&mut self, idx: u16) -> Result<NonNull<u8>, DiskError> {
+        if self.make_space_for_idx(idx)? {
+            // We do need to read, which means this wasn't already the
+            // page. Mark it as a new buffer page
+            //
+            // ELSE: we don't need a read - that means we were already there.
+            // Keep disk marked as whatever it was.
+            self.caches[0].page = PageState::Buffer(idx);
+        }
+
+        Ok(self.active_buf())
+    }
+
+    pub fn empty_buffers(&mut self) {
+        self.caches.iter_mut().for_each(|c| {
+            c.page = PageState::Empty;
+        });
+    }
+
+    pub fn mark_dirty(&mut self) {
+        let next = match self.caches[0].page {
+            PageState::Empty => {
+                // This is maybe an error?
+                PageState::Empty
+            }
+            PageState::Buffer(i) => PageState::Dirty(i),
+            PageState::Clean(i) => PageState::Dirty(i),
+            PageState::Dirty(i) => PageState::Dirty(i),
+        };
+        self.caches[0].page = next;
+    }
+
+    pub fn block(&mut self, idx: u16) -> Result<NonNull<u8>, DiskError> {
+        if self.make_space_for_idx(idx)? {
+            self.driver.read(idx, self.caches[0].buf, self.size)?;
+            self.caches[0].page = PageState::Clean(idx);
+        }
+
+        Ok(self.active_buf())
+    }
+
+    pub fn driver(&mut self) -> &mut D {
+        &mut self.driver
+    }
+
+    pub fn release(self) -> D {
+        self.driver
+    }
+}
+
+pub struct Cache {
+    buf: NonNull<u8>,
+    page: PageState,
+}
+
+impl Cache {
+    fn is_page(&self, idx: u16) -> bool {
+        let i = match self.page {
+            PageState::Empty => return false,
+            PageState::Buffer(i) => i,
+            PageState::Clean(i) => i,
+            PageState::Dirty(i) => i,
+        };
+
+        i == idx
+    }
+}
+
+#[derive(Copy, Clone, PartialEq, Debug)]
+pub enum PageState {
+    Empty,
+    Buffer(u16),
+    Clean(u16),
+    Dirty(u16),
+}
+
+#[cfg(feature = "use-std")]
+pub struct BinDisk;
+
+#[cfg(feature = "use-std")]
+impl DiskDriver for BinDisk {
+    fn read(&mut self, idx: u16, dest: NonNull<u8>, len: usize) -> Result<(), DiskError> {
+        match std::fs::read(&format!("./disk/{:05}.bin", idx)) {
```

Comment:
tiny nit, take it or leave it: IMO it would be a somewhat nicer interface if `BinDisk` could take a `PathBuf` to the location of the bin file dir, rather than always putting it in `.`?

Reply:
Yup, agreed. Will make this change.